### ISSUE TITLE
QGAN with hybrid discriminator is not being trained.

### ISSUE THREAD

**Issue by Shuhul24** [Qiskit/qiskit-machine-learning#549]
### Environment

- **Qiskit Machine Learning version**: 0.5.0
- **Python version**:  3.7
- **Operating system**: Google colab

### What is happening?

I am training a quantum generative adversarial networks where the generator is classical and the discriminator is hybrid quantum-classical network. For building such circuit, I have used the idea of hybrid quantum-classical neural network from the tutorial of qiskit [link](https://qiskit.org/documentation/machine-learning/tutorials/05_torch_connector.html#Part-2:-MNIST-Classification,-Hybrid-QNNs). But when I am training the QGAN, I am seeing no progress in the training bar, which means the network is not being trained.

### How can we reproduce the issue?

The function of quantum circuit is shown as :
```python
def create_qnn():
  feature_map = ZZFeatureMap(16)
  ansatz = RealAmplitudes(16, reps=1)
  qc = QuantumCircuit(16)
  qc.compose(feature_map, inplace=True)
  qc.compose(ansatz, inplace=True)

  qnn = EstimatorQNN(
      circuit=qc,
      input_params = feature_map.parameters,
      weight_params = ansatz.parameters,
      input_gradients=True
  )

  return qnn

qnn1 = create_qnn()
```

The discriminator network (hybrid quantum-classical neural network) is shown as:
```python
class Discriminator(nn.Module):
    def __init__(self, qnn):
        super(Discriminator, self).__init__()
        self.n_input = 784
        self.main = nn.Sequential(
            nn.Linear(self.n_input, 1024),
            nn.LeakyReLU(0.2),
            nn.Dropout(0.3),
            nn.Linear(1024, 512),
            nn.LeakyReLU(0.2),
            nn.Dropout(0.3),
            nn.Linear(512, 256),
            nn.LeakyReLU(0.2),
            nn.Dropout(0.3),
            nn.Linear(256, 16),
            TorchConnector(qnn),
            nn.Linear(1, 1),
            nn.Sigmoid(),
        )
    def forward(self, x):
        x = x.view(-1, 784)
        return self.main(x)
```

The training section of discriminator is shown as:
```python
def train_discriminator(optimizer, data_real, data_fake):
    b_size = data_real.size(0)
    real_label = label_real(b_size)
    fake_label = label_fake(b_size)
    optimizer.zero_grad()
    output_real = discriminator(data_real)
    loss_real = criterion(output_real, real_label)
    output_fake = discriminator(data_fake)
    loss_fake = criterion(output_fake, fake_label)
    loss_real.backward()
    loss_fake.backward()
    optimizer.step()
    return loss_real + loss_fake
```

### What should happen?

The network should simply train, which in my case is not happening.

### Any suggestions?

I believe that in the tutorial of QGAN [link](https://qiskit.org/documentation/machine-learning/tutorials/04_torch_qgan.html) has used hybrid quantum-classical neural network for generator, but they have also created a custom function for gradient-based training for the generator. I assume here, in my issue, as well I need to create such function, but I am not getting it somehow. Could someone help me figure out the same?

**Comment by adekusar-drl** [comment]
@Shuhul24 Sorry for the long delay, for some reasons I missed the issue. It's hard to say what's wrong from what you've posted here. In the QGAN tutorial the generator is fully quantum and the discriminator is classical. When `CircuitQNN` is in the sampling mode, it can't be used in gradient based training since it just does not return gradients in this mode. In your case you have `EstimatorQNN` that works differently and some other incomplete code. 

Back to your questions:

> I assume here, in my issue, as well I need to create such function, but I am not getting it somehow.

I believe so, and loss function depends on your setup.

> The training starts as I replace the BATCH_SIZE to 1

Why reducing batch size affected the training loop I don't know, the code is not shown here.

> have changed the number of qubits in the quantum circuit from 16 to 2. What does it mean?

You reduced the number of qubits, thus reduced the number of inputs to your network. Also, this reduced the number of parameters. Reducing the number of parameters may dramatically reduce training time and improve convergence. Is it a right decision to reduce the number of qubits? It's up to the design of the neural network. 

> Is there any limitations to TorchConnector class regards to number of qubits it can operate

No.

I'd suggest to review the QGAN tutorial again and try to update it step by step and ultimately get to the model you need.

**Comment by adekusar-drl** [comment]
@Shuhul24 Is there any help required in this issue?

**Comment by GayatriVadaparty** [comment]
There are several challenges associated with using a classical generator and a hybrid quantum-classical discriminator in a quantum generative adversarial network (QGAN).
The main advantage of using quantum computing in machine learning is its potential for speedup over classical methods. However, if the generator is classical, this advantage is lost, and the QGAN may not perform better than a classical GAN.
While classical generators are simpler to design and optimize than quantum generators, designing a generator that can generate samples from complex probability distributions can still be challenging.
A classical generator is limited by the expressiveness of the classical probability distributions it can generate. This can limit the range of possible generated samples and make it harder for the QGAN to generate high-quality samples.
The hybrid quantum-classical discriminator can be more challenging to train than a classical discriminator because it involves quantum measurements and operations. The training process may require more computational resources and be more sensitive to noise. As the size of the quantum part of the hybrid quantum-classical discriminator increases, the circuit becomes harder to simulate on classical computers. This limits the scalability of the QGAN.
In conclusion, while using a classical generator and a hybrid quantum-classical discriminator in a QGAN can be an effective approach, it also comes with some challenges that need to be addressed.

**Comment by Shuhul24** [comment]
@GayatriVadaparty thanks for the insight. Is there any mathematical implications for the above explanation that you provided?

**Comment by GayatriVadaparty** [comment]
Generally, The goal of the generator is to produce quantum states that are indistinguishable from those produced by the true data distribution, while the discriminator's goal is to correctly identify whether a given quantum state was generated by the generator or by the true data distribution. This process is repeated in an iterative manner until the generator produces quantum states that are effectively indistinguishable from the true data distribution.
So far I know, Mathematically, the generator is a unitary transformation that maps a fixed input state to a quantum state. Let U be the generator unitary, and let |0⟩ be the fixed input state. The output of the generator can be written as U|0⟩.
The discriminator, on the other hand, is a classical function that takes in the description of the quantum state produced by the generator and produces a probability that the state was generated by the generator or by the true data distribution. Let D be the discriminator function. Given a quantum state |ψ⟩, the output of the discriminator can be written as D(|ψ⟩).
The training of the QGAN involves optimizing the parameters of the generator and the discriminator. This is done by minimizing a loss function that measures the difference between the output of the discriminator for the generator-produced states and the output of the discriminator for the true data distribution. This loss function is given by:

L = −(1/N)Σᵢ[log(D(xᵢ)) + log(1−D(U|0⟩))]

where N is the number of data samples, xᵢ is a data sample, and the summation is taken over all data samples.
The goal of the optimization is to find the generator unitary U that minimizes the loss function while the discriminator D is optimized to maximize it. This process is repeated until the generator produces quantum states that are indistinguishable from those produced by the true data distribution.
I would say, the classical generator and hybrid quantum-classical discriminator work together to generate quantum states that are indistinguishable from the true data distribution, allowing for the generation of new data samples that follow the same distribution as the training data.

**Comment by Shuhul24** [comment]
Thanks for the quick reply. The explanation was very insightful and I understood somewhat a basic mathematical explanation behind the classical generator and quantum discriminator.

**Comment by adekusar-drl** [comment]
Closing this issue as stale. Feel free to re-open if you need more help.
